### ISSUE TITLE
UDP traffic loss observed during restart of service pod when externalTrafficPolicy is "local" for loadbalancer type service.

### ISSUE THREAD

**Issue by vinay50muddu** [kubernetes/kubernetes#113250]
### What happened?

Consider a UDP service of type "loadbalancer" with "externalTrafficPolicy: local" is created. When a backend pod corresponding to that service is deleted, the pod comes up in a different node. It is observed that sometimes the speaker on this new node advertises the LB IP before kube-proxy sets up the rules. The UDP traffic is redirected to the new node as the speaker on the previous node where the deleted pod was running clears it's advertisement.
Since there is no IPTABLE rule configured yet on the new node, when the first packet arrives, the packet will be dropped and a conntrack entry is created and updated with UNREPLIED. The subsequent traffic with same tuple is blackholed and updates the expiry timer until the conntrack entry is deleted. Similar case was handled by kube-proxy when "externalTrafficPolicy: cluster". But the current case where "externalTrafficPolicy: local" is not handled by the changes.

### What did you expect to happen?

The UDP stale conntrack entries on the new node are expected to be cleared. 
The current logic checks if the number of endpoints change from 0 to non-zero across the cluster and if so, marks the conntrack 
entries corresponding to those IPs (LB IP, Cluster IP, Node IP) as stale entries. These entries are removed from conntrack table.
However this logic only fixes the problem when ExternalTrafficPolicy is cluster. When ExternalTrafficPolicy is local, we need to check
if the no of endpoints on this node change from 0 to non-zero irrespective of their count across the cluster.

### How can we reproduce it (as minimally and precisely as possible)?

Create a UDP service of type "loadbalancer" with "ExternalTrafficPolicy" set to "local"
Make sure there are the replicas are more than or equal to 2.
Pump the UDP traffic(DNS client) with same ephemeral port continuosly.
Delete one pod. The pod comes up on a different node.  Check the conntrack entries. We observe that these entries are UNREPLIED and timer keeps updating.

### Kubernetes version

<details>
1.23.6
</details>

### Cloud provider

<details>
Vanilla kubernetes distribution
</details>

### OS version

<details>

SUSE linux

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by aojea** [comment]
what version are you using, there were several fixes for these scenarios

https://github.com/kubernetes/kubernetes/pull/104009

**Comment by vinay50muddu** [comment]
> what version are you using, there were several fixes for these scenarios
> 
> #104009

Thanks @aojea for the reply. We are using 1.23.6 as mentioned earlier. I have gone through the link you referred to. The fix works when externalTrafficPolicy=cluster. Because the logic checks if number of endpoints in the cluster is changed from 0 to non-zero and if so marks the entries as stale. But when the externalTrafficPolicy=local this may not work. Because, in this case we only care about endpoints on the current node whose count change from 0 to non-zero (Please note that there may be endpoints existing in the cluster  even before but not on this node).

**Comment by aojea** [comment]
> Since there is no IPTABLE rule configured yet on the new node, when the first packet arrives, the packet will be dropped and a conntrack entry is created and updated with UNREPLIED

there should be an iptables rule to drop , at least since 1.25.
However, can you share more details about the software that you are using for BGP and loadbalancer?
It should not direct traffic to a Node without endpoints :/

**Comment by vinay50muddu** [comment]
> > Since there is no IPTABLE rule configured yet on the new node, when the first packet arrives, the packet will be dropped and a conntrack entry is created and updated with UNREPLIED
> 
> there should be an iptables rule to drop , at least since 1.25. However, can you share more details about the software that you are using for BGP and loadbalancer? It should not direct traffic to a Node without endpoints :/

There is a rule to drop. So the packet is UNREPLIED and registered in conntrack. I didn't understand your comment " It should not direct traffic to a Node without endpoints :/". There is a new endpoint now on this node but just that the rules are not configured.
We are using old version of metallb controller and ECFE k8s agent that uses BIRD speaker for BGP. 
BIRD version: v2.0.6
metallb controller version:  v0.12.1

**Comment by aojea** [comment]
> . I didn't understand your comment

I got it wrong , sorry, I understand now the scenario

**Comment by aojea** [comment]
with ` externalTrafficPolicy=local`the kube-proxy exposes a healthcheck to indicate if there are endpoints ready (as seen by the proxy) on that node, is this metallb implementation checking this endpoint healthcheck before sending traffic to that node?

**Comment by adilGhaffarDev** [comment]
Tested it and have been able to reproduce the issue. Created a simple udp server deployment with udp service attached to it. In UDP service externalTrafficPolicy is set to local. There are 2 replicas of the deployment  in different nodes, to reproduce the scenerio I added label on node and also added nodeSelector with this label in the deployment yaml, both replicas moved to one node and UNREPLIED conntrack entries can be seen:
```
udp      17 11 src=192.168.49.1 dst=192.168.49.241 sport=57474 dport=10001 [UNREPLIED] src=192.168.49.241 dst=192.168.49.1 sport=10001 dport=57474 mark=0 use=1
udp      17 5 src=192.168.49.1 dst=192.168.49.241 sport=33837 dport=10001 [UNREPLIED] src=192.168.49.241 dst=192.168.49.1 sport=10001 dport=33837 mark=0 use=1
udp      17 1 src=192.168.49.1 dst=192.168.49.241 sport=47461 dport=10001 [UNREPLIED] src=192.168.49.241 dst=192.168.49.1 sport=10001 dport=47461 mark=0 use=1
udp      17 13 src=192.168.49.1 dst=192.168.49.241 sport=42775 dport=10001 [UNREPLIED] src=192.168.49.241 dst=192.168.49.1 sport=10001 dport=42775 mark=0 use=1
udp      17 9 src=192.168.49.1 dst=192.168.49.241 sport=51096 dport=10001 [UNREPLIED] src=192.168.49.241 dst=192.168.49.1 sport=10001 dport=51096 mark=0 use=1
udp      17 3 src=192.168.49.1 dst=192.168.49.241 sport=35713 dport=10001 [UNREPLIED] src=192.168.49.241 dst=192.168.49.1 sport=10001 dport=35713 mark=0 use=1
udp      17 7 src=192.168.49.1 dst=192.168.49.241 sport=60395 dport=10001 [UNREPLIED] src=192.168.49.241 dst=192.168.49.1 sport=10001 dport=60395 mark=0 use=1
```

192.168.49.241 is the ip of udp service. I have created kube-proxy image with the solution and tested it, it fixes the issue. I can assign this to myself and create PR, if its ok.

**Comment by aojea** [comment]
but those unreplied entries eventually will timeout and disappear, the point is that with `externalTrafficPolicy` the node that has no endpoint should not receive traffic

Because there may be some edge cases I may be missing , if you are able to reproduce the bug with an e2e test (the bug is traffic should be blackholed not that conntrack entries take time to be deleted) we have all the related tests here https://github.com/kubernetes/kubernetes/blob/master/test/e2e/network/conntrack.go , feel free to send a PR and tag me and we follow up there

**Comment by adilGhaffarDev** [comment]
@aojea just to to confirm I have run few more tests and traffic is being blackholed. I will try to reproduce the issue with e2e and give you the update.

**Comment by thockin** [comment]
Any updates here?

**Comment by uablrek** [comment]
**TL;DR;** Please test with the latest K8s.

Same setup as in https://github.com/kubernetes/kubernetes/issues/113203#issuecomment-1342547979.

### Test case

1. Start continuous traffic for 30s, and log statistics
2. Wait 5s
3. Re-route traffic to the new node
4. Wait 1s, then collect conntrack data on the new node
5. Move the POD to the new node
6. Wait for the traffic generation to stop
7. Collect statistics

Some traffic is allowed to the new node and then `conntrack -L` is executed.

## Result

UDP v1.23.6:
<img src="https://user-images.githubusercontent.com/37046727/206509121-c924c970-6153-4139-ba19-fc0f155f08e2.svg" width="60%" />

UDP v1.25.3:
<img src="https://user-images.githubusercontent.com/37046727/206509889-9b41bbed-b96c-45c2-bdb8-8151cb8e2c53.svg" width="60%" />

With v1.23.6 conntrack entries are seen on the new node when traffic is re-directed, but the POD is not started:
```
udp      17 29 src=192.168.1.201 dst=10.0.0.52 sport=38555 dport=5003 [UNREPLIED] src=10.0.0.52 dst=192.168.1.201 sport=5003 dport=38555 mark=0 use=1
udp      17 29 src=192.168.1.201 dst=10.0.0.52 sport=37207 dport=5003 [UNREPLIED] src=10.0.0.52 dst=192.168.1.201 sport=5003 dport=37207 mark=0 use=1
udp      17 29 src=192.168.1.201 dst=10.0.0.52 sport=37398 dport=5003 [UNREPLIED] src=10.0.0.52 dst=192.168.1.201 sport=5003 dport=37398 mark=0 use=1
udp      17 29 src=192.168.1.201 dst=10.0.0.52 sport=52864 dport=5003 [UNREPLIED] src=10.0.0.52 dst=192.168.1.201 sport=5003 dport=52864 mark=0 use=1
udp      17 29 src=192.168.1.201 dst=10.0.0.52 sport=48892 dport=5003 [UNREPLIED] src=10.0.0.52 dst=192.168.1.201 sport=5003 dport=48892 mark=0 use=1
udp      17 29 src=192.168.1.201 dst=10.0.0.52 sport=46212 dport=5003 [UNREPLIED] src=10.0.0.52 dst=192.168.1.201 sport=5003 dport=46212 mark=0 use=1
udp      17 29 src=192.168.1.201 dst=10.0.0.52 sport=46738 dport=5003 [UNREPLIED] src=10.0.0.52 dst=192.168.1.201 sport=5003 dport=46738 mark=0 use=1
udp      17 29 src=192.168.1.201 dst=10.0.0.52 sport=43114 dport=5003 [UNREPLIED] src=10.0.0.52 dst=192.168.1.201 sport=5003 dport=43114 mark=0 use=1
udp      17 29 src=192.168.1.201 dst=10.0.0.52 sport=42839 dport=5003 [UNREPLIED] src=10.0.0.52 dst=192.168.1.201 sport=5003 dport=42839 mark=0 use=1
udp      17 29 src=192.168.1.201 dst=10.0.0.52 sport=47899 dport=5003 [UNREPLIED] src=10.0.0.52 dst=192.168.1.201 sport=5003 dport=47899 mark=0 use=1
```
but with v1.25.3 there are none.

**Comment by uablrek** [comment]
> Pump the UDP traffic(DNS client) with same ephemeral port continuosly

I think my test program select new ephemeral ports. So my test is not as described.

However, the conntrack clean-up has seen some updates since v1.23.6, so please test with the latest K8s anyway so we don't hunt bugs that are solved.

**Comment by thockin** [comment]
Ping for updates?

**Comment by adilGhaffarDev** [comment]
> > Pump the UDP traffic(DNS client) with same ephemeral port continuosly
> 
> I think my test program select new ephemeral ports. So my test is not the same as described in the issue.
> 
> However, the conntrack clean-up has seen some updates since v1.23.6, so please test with the latest K8s anyway so we don't hunt bugs that are solved.

Thank you @uablrek  for your investigation, I have tested it too in both v1.23.6 and v1.25.3 with the same ephemeral port, but I was not able to reproduce the issue in the latest version and it is reproducible in the older version. Also with daemonset and just changing nodename I was seeing unreplied entries but I wasn't able to reproduce the issue where traffic is blackholed and unreplied conntrack entries are stuck, I had to change my setup a little bit as below:

1. Create loadbalancer svc with ETP local and external IP
2. Added nodeselector in UDP server deployment and same label on node yaml.
3. Create 2 replicas of deployment, both will schedule on the same node n1.
4. Add static route to the node(n1) where pods are running
5. Pump UDP traffic with the same ephemeral port
6. Now remove the label from node n1 and set it on the other one n2 where no pod is running.
7. Change the static route to n2
8. Delete one of the pod, it will be rescheduled on n2:
   a. In the case of older k8s version conntrack does not update and traffic is blackholed. Unreplied entry stuck not timing out.
   b. In the case of newer version conntrack entries are updated and traffic can be seen.

**Comment by uablrek** [comment]
@adilGhaffarDev So, can we close this issue as "magically fixed in newer K8s versions"?

**Comment by adilGhaffarDev** [comment]
> @adilGhaffarDev So, can we close this issue as "magically fixed in newer K8s versions"?

If @vinay50muddu is ok with that, we can close this ticket.

**Comment by uablrek** [comment]
@adilGhaffarDev https://github.com/kubernetes/kubernetes/issues/113203 seems very similar. If possible, can you please take a look at that one too?
